Add doc_count field mapper

csoulios opened the pull request:
Bucket aggregations compute bucket `doc_count` values by incrementing the `doc_count` by 1 for every document collected in the bucket. 

When using summary fields (such as `aggregate_metric_double`) one field may represent more than one document. To provide this functionality we have implemented a new field mapper (named `doc_count` field mapper). This field is a positive integer representing the number of documents aggregated in a single summary field.

Bucket aggregations will check if a field of type `doc_count` exists in a document and will take this value into consideration when computing doc counts.

csoulios committed:
Initial version of doc_count field mapper

csoulios committed:
added tests

csoulios committed:
Build fixes

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Added tests for doc_count fieldmapper

csoulios committed:
doc count tests

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Resolve conflicts after merge from master

csoulios committed:
Added yaml test for doc_count field type

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Minor changes to test

csoulios committed:
Fix issue with not-registering field mapper

csoulios committed:
Simplify terms agg test

elasticmachine commented:
Pinging @elastic/es-analytics-geo (:Analytics/Aggregations)

elasticmachine commented:
Pinging @elastic/es-search (:Search/Mapping)

csoulios committed:
Add doc_count provider in the buckets aggregator

jtibshirani reviewed (commented):
Should we make this a metadata field `_doc_count` instead of a standard field type? It could be a nice fit, because it's more of an internal field rather than something we expect users to configure. Also we don't want multiple fields of type `doc_count` to be defined.

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios commented:
@jtibshirani I was planning to add a validation that does not allow more than one `doc_count` fields in the mapping. 

Also, I am not sure what are the properties or limitations of metadata fields that would affect this use case. For example, can users set or modify the value of a metadata field? Guidance on this topic is very welcome.

jtibshirani commented:
> For example, can users set or modify the value of a metadata field? Guidance on this topic is very welcome.

Would you be able to clarify what you mean by 'set or modify the value'?

talevy commented:
I believe the reason it is not a metadata field is because it is intended to be set explicitly by the user. It is meant to be used in tandem with the aggregate metric fields.

Some additional questions I have is surrounding how it is expected to be used by users of the histogram field. Is it expected that users like [apm-server](https://github.com/elastic/apm-server/issues/3485) add a `doc_count` field in addition to the histogram?

csoulios commented:
With regards to `histogram` fields, behavior should be similar to the `aggregate_metric_double` field. `histogram` field (just like the `aggregate_metric_double` field) stores the summary values, while  the `doc_count` field in the document  represents the number of documents aggregated to produce those summary values.

For example:
```
{
  "latency": {
    "values" : [0.1, 0.2, 0.3, 0.4, 0.5], 
    "counts" : [3, 7, 23, 12, 6]
  }
  "cpu_load": {
    "min": 3,
    "max": 95
  },
  "doc_count": 100
}
```

In the document above, we aggregated 100 "raw" events and produced a histogram of the `latency` field and a `min`/`max` summary of the `cpu_load` field.

polyfractal commented on server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java:
```diff
@@ -0,0 +1,159 @@
+    public static final class DocCountFieldType extends MappedFieldType {
+
+        public DocCountFieldType(String name, boolean hasDocValues, Map<String, String> meta) {
+            super(name, false, hasDocValues, TextSearchInfo.SIMPLE_MATCH_ONLY, meta);
```
Needs verification from someone that knows mappers better than me, but I think if we set `IndexOptions.NONE` we should probably also set `TextSearchInfo.NONE`.  I think?  Not entirely confident in that

csoulios replied:
Class `VersionFieldMapper` uses similar pattern (`IndexOptions.NONE` and `TextSearchInfo.NONE`)

https://github.com/elastic/elasticsearch/blob/62f51eb9aec6c3d423d7fa9645ebe9ef2904d4c1/server/src/main/java/org/elasticsearch/index/mapper/VersionFieldMapper.java#L75

I am not an expert on this either. 

@jtibshirani could you help with this please?

jtibshirani replied:
I also think this should be `TextSearchInfo.NONE`, given we only allow `IndexOptions.NONE`. The choice in `VersionFieldMapper` seems like a mistake -- `TextSearchInfo` was created recently and we probably this during review.

csoulios replied:
Sure, happy to change that to `TextSearchInfo.NONE`

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java:
```diff
@@ -392,4 +397,21 @@ public static boolean descendsFromGlobalAggregator(Aggregator parent) {
     }
 
+    @Override
+    protected void preGetSubLeafCollectors(LeafReaderContext ctx) throws IOException {
```
I think we should move this to `doPreCollection()` if possible.  That will get called once before starting to collect docs on the entire shard, whereas `preGetSubLeafCollectors()` is called once per segment in the shard

If it's not possible to move for some reason, we should add a `if (fieldDocCountProvider != null) { return } ` check so we only load it once

csoulios replied:
`FieldBasedDocCountProvider` requires the instance of the `LeafReaderContext` for reading the `doc_count` field. 

In commit c0f23ae I separated the code that resolves the `DocCountProvider` (and put this part in the constructor) from loading the `LeafReaderContext` and put it in the `preGetSubLeafCollectors()` method.

polyfractal replied:
Looks great, I like this modification a lot! :+1:   I think the extra method on the interface is worth it in this scenario

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/FieldDocCountProvider.java:
```diff
@@ -0,0 +1,52 @@
+
+/**
+ * An implementation of a doc_count provider that reads the doc_count value
+ * in a doc value field in the document
```
Could we add a note that if the doc doesn't have the doc_count field, it will return 1 as the count?

polyfractal reviewed (changes requested):
Left some comments!  You were right, it does have a pretty minimal impact on the aggs overall, nice! :)

I wonder if we should enforce documents have the doc_count field if the mapping includes it?  Not sure if we have a good way to enforce that though, and it would be very opposite of how other ES fields work.  Don't feel strongly about it, just a thought.

Having the default of 1 for documents without a count seems reasonable, we should make sure we document it.

Speaking of, do we need docs for this?

csoulios commented on server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java:
```diff
@@ -0,0 +1,159 @@
+
+        static {
+            FIELD_TYPE.setDocValuesType(DocValuesType.NUMERIC);
+            FIELD_TYPE.setIndexOptions(IndexOptions.NONE);
```
I was thinking about the same thing and I could not find a good reason to make this field indexed. For the sake of simplicity and to avoid indexing information not likely to be queried, I decided not to. 

Nevertheless, I don't have a strong opinion on this, so I am open to suggestions.

polyfractal replied:
Yeah I can't think of a good reason now either.  We can always enable it later if we want it.

Alternatively, we could override `public Builder indexOptions(IndexOptions indexOptions) {...}` and allow the user to enable indexing on the field, but leave it disabled by default.

csoulios committed:
Initialize doc_count provider once

doc_count provider is initialized once in the
buckets agg constructor.

csoulios commented:
> I wonder if we should enforce documents have the doc_count field if the mapping includes it? Not sure if we have a good way to enforce that though, and it would be very opposite of how other ES fields work. Don't feel strongly about it, just a thought.

I think that we should make this lenient and flexible. We should not require a `doc_count` field for every document. If the field is missing we fallback to the default behavior that is `doc_count = 1`. And yes, we must document this.

csoulios commented:
In our initial conversations about the `doc_count` field type we had discussed that if an index contains an `aggregate_double_metric` field (or any other summary field) we should require that there is a `doc_count` field. Now, I don't think we should be so strict. Instead, we should give the option to the user. I suppose there are use cases that require summary fields but don't actually care about doc counts. For those cases, users will have to add a `doc_count` field that will always contain no value. 

@talevy @polyfractal WDYT?

csoulios committed:
Added tests for FieldBasedDocCountProvider

csoulios committed:
Added more tests to DocCountFieldMapper

csoulios committed:
Fixed NPE at AggregatorTestCase

polyfractal commented:
@csoulios Yeah I think you're right, we should be lenient here.  APM for example is happily using `histogram` right now without `doc_count`.  They may end up adding `doc_count`, but if they are happy with the functionality right now it doesn't make sense to break their workflow with a strict requirement.

++ to leaving it lenient, and also ++ to not requiring it in every document

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Updated branch to fix build after merge

csoulios committed:
Added validation for single doc_count field

Added validation to ensure only a single doc_count
field exists in the mapping

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc

csoulios committed:
Added version skips to fix broken tests

csoulios committed:
Added documentation for doc_count

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

csoulios committed:
Changes to address review comments:

- Minor doc change
- Added yml test that merges template with multiple doc_count
- Changed DocCountFieldType indexing to TextSearchInfo.NONE

jimczi commented on server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java:
```diff
@@ -0,0 +1,160 @@
+                    "Field [" + fieldType().name() + "] must be a positive integer.");
+            }
+
+            final Field docCount = new NumericDocValuesField(name(), value.longValue());
```
I wonder if you can use a canonical name for the Lucene field since we allow a single doc_count field per mapping. Something like ` _doc_count` would be easier to manipulate rather than looking in the mapping if a doc_count field exists for every request ?

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java:
```diff
@@ -63,6 +67,16 @@ public BucketsAggregator(String name, AggregatorFactories factories, SearchConte
+        for (MappedFieldType fieldType : context.getQueryShardContext().getMapperService().fieldTypes()) {
+            if (DocCountFieldMapper.CONTENT_TYPE.equals(fieldType.typeName())) {
+                // If a field of type doc_count has been found, use it to provide the bucket doc_count values
+                fieldDocCountProvider = new FieldBasedDocCountProvider(fieldType.name());
```
Using a canonical name for the Lucene field could simplify this part ? We could use the `FieldBasedDocCountProvider` all the time and the logic to handle missing values/field would be implemented directly in this concrete class (no need to introduce an interface) ?

csoulios replied:
With the interface, I tried to have a more generic approach. Since we are moving to the concept of a summary "document" maybe in the future we will need more implementations. For example, extract `doc_count` from a script. Another case would be to use the `histogram` buckets count as the `doc_count`. 

Also, in 99% of the cases the good old implementation of incrementing by 1 is what we need. In this  case, reading fields and handling missing values would be more overhead. Resolving the correct `DocCountProvider` instance once in the `SearchContext` is minimal overhead imho.

jimczi replied:
> Also, in 99% of the cases the good old implementation of incrementing by 1 is what we need. In this case, reading fields and handling missing values would be more overhead. 

We'd just need to check if the `_doc_count` field exists in the Lucene segment ? If not, then we can fallback to return 1 all the  time. My point here is that it would simplify things to expect a canonical field name in Lucene.

csoulios replied:
I pushed commit 39c43a0. 

`DocCountFieldMapper` uses canonical name `_doc_count` for the Lucene field and I modified `FieldBasedDocCountProvider` to only look for this field. It already had the logic to handle missing values/field in the Lucene segment. 

As suggested, I modified `BucketAggregator` to use `FieldBasedDocCountProvider`  by default, without looking up the index mapping for a `doc_count` field. 

So, I have the following questions: 

1. Is there a significant performance overhead for looking up a field for each lucene segment? Especially, when 99% of the times the `_doc_count` field is missing?

2. If performance overhead is negligible, does it still make sense to instantiate `FieldBasedDocCountProvider` in the `SearchContext`? I hesitate leaking this implementation aspect of the `BucketAggregator` to the `SearchContext`.

jimczi replied:
> 1

I think it's ok and will be sometimes better than checking every field in the mapping.

> 2

I agree that it's easier to create a new iterator for each multi-bucket aggregator. Maybe that's the best approach to start with since accessing these values should be fast. Although, if we keep the current design, I think we can use a plain `NumericDocValues` rather than introducing the `DocCountProvider` interface.

csoulios replied:
> I think it's ok and will be sometimes better than checking every field in the mapping.

Agreed. I will go down that road. Although the initial implementation would check mapping once per shard and not once per segment. 

> Although, if we keep the current design, I think we can use a plain NumericDocValues rather than introducing the DocCountProvider interface.

For the sake of the separation of concerns principle, I would keep the `FieldBasedDocCountProvider` instead of using bare bones `NumericDocValues`. I think it makes code cleaner and more testable. Also, this is a class we can pull up at `SearchContext` if we want at a later stage. If you feel that there's too much abstraction, I can remove the `DocCountProvider` interface and only keep the concrete class.

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java:
```diff
@@ -91,7 +105,8 @@ public final void collectBucket(LeafBucketCollector subCollector, int doc, long
     public final void collectExistingBucket(LeafBucketCollector subCollector, int doc, long bucketOrd) throws IOException {
-        if (docCounts.increment(bucketOrd, 1) == 1) {
+        int docCount = fieldDocCountProvider.getDocCount(doc);
+        if (docCounts.increment(bucketOrd, docCount) == 1) {
```
Should it be changed to `docCounts.increment(bucketOrd, docCount) == docCount` ?

csoulios replied:
Yes, it seems that I missed it. Thanks

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java:
```diff
@@ -52,6 +55,7 @@
     private final BigArrays bigArrays;
     private final IntConsumer multiBucketConsumer;
     private IntArray docCounts;
+    private DocCountProvider fieldDocCountProvider = DocCountProvider.DEFAULT_PROVIDER;
```
I wonder if we can have a single instance of this provider per request ? Maybe it should be added in the `SearchContext` and re-initiated every time we need to perform a collection ? With @nik9000 refactoring of the breadth first execution I think it should be easier to plug the logic on a higher level.

jimczi reviewed (changes requested):
I left some comments/questions. I like the approach but I think we should avoid duplicating the doc_count provider per multi-bucket aggregation.

csoulios committed:
Use _doc_count as Lucene field for doc count

csoulios committed:
Minor change: field rename

jimczi commented:
> Should we make this a metadata field _doc_count instead of a standard field type? It could be a nice fit, because it's more of an internal field rather than something we expect users to configure. Also we don't want multiple fields of type doc_count to be defined.

We don't allow metadata field in the `_source` today but I think this is something we should discuss since it seems more natural to define this new field as a metadata field. I am not even sure that it should be disabled by default, we could add an option to make it required like `_routing`:

```yaml
"_doc_count": {
      "required": true 
}
```
 ... but otherwise it should be possible to override the implicit 1 by providing an explicit `_doc_count` when indexing a document ?
I don't have strong feelings regarding the presence of metadata fields in `_source`, it would be nice to be able to provide this information outside of `_source` (like existing metadata field) but that's a much bigger task so I wonder what others think. @jtibshirani is it what you had in mind when you proposed to use a metadata field ?

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc

jtibshirani commented:
> @jtibshirani is it what you had in mind when you proposed to use a metadata field ?

@jimczi yes this is exactly what I had in mind. I like the idea of having it enabled by default, with a `required` option.

> I don't have strong feelings regarding the presence of metadata fields in _source, it would be nice to be able to provide this information outside of _source (like existing metadata field) but that's a much bigger task so I wonder what others think.

I would also be okay allowing metadata field values to be specified in `_source`. I don't see any reasons why we couldn't support this. @csoulios for context, would you be able to explain how we expect rollups to specify this value? I guess rollups just create a new field in the `_source` representing the doc count?

csoulios commented:
>  @csoulios for context, would you be able to explain how we expect rollups to specify this value? I guess rollups just create a new field in the `_source` representing the doc count?

@jtibshirani, we want to allow users to set this value per document. For example allow the following request:
```
PUT /rollup-index/_doc/1
{
  "cpu_load": {
    "min": 3,
    "max": 95
  },
  "_doc_count": 100
}
```

Also, please have in mind that using a `_doc_count` field can be used outside rollups. There can be cases that users may want to use summary fields and setting `_doc_count`s without this being a part of rollups.

jtibshirani commented:
> Also, please have in mind that using a _doc_count field can be used outside rollups

That makes sense. I was actually asking about rollups in particular, whether they need a special way to pass through the value. Based on your answer, it sounds like rollups will just add the value to `_source` for now.

csoulios committed:
Minor change to yml test.

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Fix errors from merge

csoulios committed:
Converted _doc_count to metadata field type

csoulios committed:
Throw an error if parsed value is not a number

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Make _doc_count field a metadata field

csoulios committed:
Fixed broken tests

csoulios committed:
Fix bug in low cardinality ordinal terms aggs

csoulios committed:
Update docs that _doc_count is a metadata field

csoulios committed:
Fix broken ML tests

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Fix errors after merge

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java:
```diff
@@ -53,6 +54,7 @@
     private final BigArrays bigArrays;
     private final IntConsumer multiBucketConsumer;
     private IntArray docCounts;
+    protected final DocCountProvider docCountProvider;
```
We could use a `NumericDocValues` directly and have the logic in a protected function like `getDocCountForDoc(int)` ? We don't really need the new interface and classes ?

csoulios replied:
Method `DocCountProvider#getDocCount(int)` is used by `BucketsAggregator` and `SortedDocsProducer` classes

https://github.com/elastic/elasticsearch/blob/7b7ca43d6aa9bac621f3977c519a407122e10623/server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java#L98

https://github.com/elastic/elasticsearch/blob/7b7ca43d6aa9bac621f3977c519a407122e10623/server/src/main/java/org/elasticsearch/search/aggregations/bucket/composite/SortedDocsProducer.java#L72

I think the easier way to share the code of reading `_doc_count` field is to encapsulate it in a small simple class and call it from both `BucketsAggregator` and `SortedDocsProducer` classes. To simplify the class even more, I have modified `DocCountProvider` interface and made it a concrete class with the implementation.

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/BucketsAggregator.java:
```diff
@@ -53,6 +54,7 @@
     private final BigArrays bigArrays;
     private final IntConsumer multiBucketConsumer;
     private IntArray docCounts;
```
Should we switch to long now that a single document can add more than 1 ?
That's probably a follow up though, no need to change in this PR.

csoulios replied:
Fixed in this PR

jimczi commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java:
```diff
@@ -305,6 +305,7 @@ public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCol
             assert sub == LeafBucketCollector.NO_OP_COLLECTOR;
             final SortedDocValues singleValues = DocValues.unwrapSingleton(segmentOrds);
             mapping = valuesSource.globalOrdinalsMapping(ctx);
+            docCountProvider.setLeafReaderContext(ctx);
```
That shouldn't be needed since `preGetSubLeafCollectors` is already called ?

jimczi reviewed (commented):
I left some comments. I like the fact that the `_doc_count` metadata field is always activated and can be leveraged on any indices. I think the `composite` aggregation needs to be changed too, see `CompositeValuesCollectorQueue`.

csoulios committed:
Addressed review comments

polyfractal commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/GlobalOrdinalsStringTermsAggregator.java:
```diff
@@ -305,6 +305,7 @@ public LeafBucketCollector getLeafCollector(LeafReaderContext ctx, LeafBucketCol
             assert sub == LeafBucketCollector.NO_OP_COLLECTOR;
             final SortedDocValues singleValues = DocValues.unwrapSingleton(segmentOrds);
             mapping = valuesSource.globalOrdinalsMapping(ctx);
+//            docCountProvider.setLeafReaderContext(ctx);
```
leftover?

polyfractal commented on server/src/test/java/org/elasticsearch/search/aggregations/bucket/FieldBasedDocCountProviderTests.java:
```diff
@@ -0,0 +1,108 @@
+import static java.util.Collections.singleton;
+
+
+public class FieldBasedDocCountProviderTests extends AggregatorTestCase {
```
Is it possible to filter or aggregate on the `_doc_counts` field itself...I'm guessing it's not possible?  Should we add some tests to make sure that's handled politely if a user tries to search/agg it?

jtibshirani replied:
+1, it's common to have a test case like `DocCountFieldTypeTests` that checks query and field data construction works (or throws an error) as expected.

csoulios replied:
I have added test `DocCountFieldTypeTests` that tests `range`, `term` and `exists` queries. 

Also, `DocCountFieldType` extends the `MappedFieldType` class that by default throws an exception when calling the `fielddataBuilder()`

polyfractal reviewed (approved):
LGTM, although my mapper-skills are relatively poor so take it with a grain of salt. :)

Left a few comments about validation/tests but not critical (and might not be relevant).  Exciting!

jtibshirani commented on docs/reference/mapping/fields.asciidoc:
```diff
@@ -76,3 +81,4 @@ include::fields/source-field.asciidoc[]
 
 include::fields/type-field.asciidoc[]
 
+include::fields/doc-count-field.asciidoc[]
```
Really small comment, I think we try to keep these alphabetized.

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java:
```diff
@@ -0,0 +1,132 @@
+
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        if (context.parser().currentToken() == XContentParser.Token.VALUE_NUMBER) {
```
Small comment, just nice to know about -- it's common to use `XContentParserUtils.ensureExpectedToken` instead of checking this manually.

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java:
```diff
@@ -0,0 +1,132 @@
+    @Override
+    protected void parseCreateField(ParseContext context) throws IOException {
+        if (context.parser().currentToken() == XContentParser.Token.VALUE_NUMBER) {
+            Long value = context.parser().longValue(false);
```
I think this could just be a primitive `long`, and that we could remove the null check and unboxing below.

Also I like that we avoid parsing leniency (`coerce: false`), something I hope we do more of going forward !

jtibshirani commented on server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java:
```diff
@@ -0,0 +1,132 @@
+    public void preParse(ParseContext context) { }
+
+    @Override
+    public ValueFetcher valueFetcher(MapperService mapperService, SearchLookup lookup, String format) {
```
We **just** moved this method to `MappedFieldType`, so this will have to be moved to `DocCountFieldType` before merging. Also it would be great to add a test -- the most common place to do that is in a field type test like `BooleanFieldTypeTests#testFetchSourceValue`.

jtibshirani reviewed (commented):
It looks good to me overall, I left some minor comments.

csoulios committed:
Addressed reviewer comments

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Added DocCountFieldTypeTests

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Fix errors after merge

csoulios committed:
Make composite agg respect _doc_count field

Cleaned up/simplified DocCountProvider class

csoulios commented on docs/reference/mapping/fields/doc-count-field.asciidoc:
```diff
@@ -0,0 +1,118 @@
+
+[IMPORTANT]
+========
+* A `_doc_count` field can only store a single positive integer per document. Nested arrays are not allowed.
```
There is an assertion that the inpurt is a `VALUE_NUMBER` in the `parseCreateField()` method.

https://github.com/elastic/elasticsearch/blob/7b7ca43d6aa9bac621f3977c519a407122e10623/server/src/main/java/org/elasticsearch/index/mapper/DocCountFieldMapper.java#L113

csoulios commented:
@elasticmachine run elasticsearch-ci/bwc
@elasticmachine run elasticsearch-ci/default-distro

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

jtibshirani commented on server/src/main/java/org/elasticsearch/search/aggregations/bucket/DocCountProvider.java:
```diff
@@ -0,0 +1,53 @@
+        try {
+            docCountValues = DocValues.getNumeric(ctx.reader(), DocCountFieldMapper.NAME);
+        } catch (IOException e) {
+            docCountValues = null;
```
I wonder if we should be swallowing an `IOException` and treating it as if there was no doc count field. What case does this logic address?

csoulios replied:
Initially, I felt it was better to swallow the error and fallback to `doc_count=1`. Now, it feels more transparent to rethrow the exception.

I am not sure what can go wrong there and I don't have a strong opinion about this.

jtibshirani replied:
I agree we should propagate it, an `IOException` could indicate something serious and it's best not to silently continue on. Perhaps this method could just declare `throws IOException`.

jtibshirani reviewed (commented):
Thanks for the updates, I had one last comment.

csoulios committed:
DocCountProvider rethrows IOException instead of swallowing it

jimczi reviewed (approved):
LGTM too.

csoulios committed:
Set familyTypeName of _doc_count to integer

csoulios committed:
Merge branch 'feature/aggregate-metrics' into doc_count-field-mapper

csoulios commented:
This PR is closed and opened as new PR (#64503) against master branch.
